unify native MIDI modules

rfomin opened the pull request:
Fix #1209 

Alternatively, we can use the PortMidi library.

rfomin commented:
Works on Linux with Nuked-SC55 emu, although the device name looks odd:
![woof0000](https://github.com/fabiangreffrath/woof/assets/5077629/12c003e7-5f59-41ba-b0be-414ad45f6c24)

rfomin commented on src/i_midimusic.c:
```diff
@@ -138,273 +123,82 @@ typedef struct
+// static void SendNOPMsg(unsigned int delta_time)
+// {
+//     ;
+// }
```
@ceski-1 I commented out `SendNOPMsg`, do we need this function?

ceski-1 replied:
It was used to preserve timing for events, e.g. looping correctly at the end of a track. I'm not sure if it's needed with the new library (I haven't reviewed or tested the PR).

rfomin replied:
Now we send messages directly and manage the time ourselves. I vaguely remember looping problems, do you have test files?

ceski-1 replied:
I'm almost certain the following will break: emidi songs, complex songs when using vanilla midi complevel, songs with loop points. I won't be able to do a thorough review for a while, sorry.

rfomin replied:
I tested some MIDI's from FF VII (XG, Yamaha S-YXG50) and Shadow Warrior (EMIDI, MS GS Wavetable Synth). It seems to work fine, but I might be missing some details.

It shouldn't be too bad, I tried to preserve the logic.

fabiangreffrath commented:
> Works on Linux with Nuked-SC55 emu, although the device name looks odd:

I think you can strip out the part outside the colons.

https://mido.readthedocs.io/en/latest/backends/rtmidi.html#port-names-linux-alsa

fabiangreffrath commented:
> > Works on Linux with Nuked-SC55 emu, although the device name looks odd:
> 
> I think you can strip out the part outside the colons.
> 
> https://mido.readthedocs.io/en/latest/backends/rtmidi.html#port-names-linux-alsa

Here's the code that explains how the string is composed:

https://github.com/thestk/rtmidi/blob/e8c8791823ca99956adb4faf2b28bb843c7c1f6e/RtMidi.cpp#L2447-L2455

rfomin commented:
> Here's the code that explains how the string is composed:

Thanks, I'm just not sure how ports work in Linux. In Windows, the device name without the port number is sufficient. Maybe we can just store the port number in the config and partially display the name in the menu.

fabiangreffrath commented:
> Thanks, I'm just not sure how ports work in Linux. In Windows, the device name without the port number is sufficient. Maybe we can just store the port number in the config and partially display the name in the menu.

I'm not sure, either. I could live with the constructed string, but if we want to make it clearer, we could probably just terminate it at the first colon. 🤷

fabiangreffrath commented:
This is what it says if I started fluidsynth as a daemon in the background before starting the engine:

![woof0000](https://github.com/fabiangreffrath/woof/assets/6449237/3c2a9abf-43f5-4fa2-9869-d55aa69e051a)

The complete string is supposed to read
```
FLUID Synth (8072):Synth input port (8072:0) 128:0
```

I think we can terminate strings at the first `':'` on Linux.

Also, this should *not* be the default choice on Linux, because without a daemon such as fluidsynth running in the background (or an actual hardware MIDI player implementation), there won't be any music at all.

fabiangreffrath commented:
And this is the string for timidity:
```
TiMidity:TiMidity port 0 128:0
```

rfomin commented:
The latest version of the Nuked SC-55 now has a name:
![woof0000](https://github.com/fabiangreffrath/woof/assets/5077629/8c4c1203-cd69-49a6-95ed-533539289713)

> Also, this should not be the default choice on Linux, because without a daemon such as fluidsynth running in the background (or an actual hardware MIDI player implementation), there won't be any music at all.

Agreed, will do some refactoring in i_sound.c later.

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -1310,19 +1311,40 @@ static void GetDevices(void)
-    CheckError("GetDevices");
+        if (current_api == RTMIDI_API_LINUX_ALSA)
+        {
+            char *p = strchr(name, ':');
```
I think this breaks when the name is not unique. A simple example is Timidity which has a few ports open by default. If port 2 is selected, the wrong port is used the next time Woof is launched (always goes back to the first found port I think).
```
$ aplaymidi -l
 Port    Client name                      Port name
 14:0    Midi Through                     Midi Through Port-0
128:0    TiMidity                         TiMidity port 0
128:1    TiMidity                         TiMidity port 1
128:2    TiMidity                         TiMidity port 2
128:3    TiMidity                         TiMidity port 3
```

rfomin replied:
We can save the full port name in config, but display only the client name and port number in the menu.

rfomin replied:
Fixed by switching to PortMidi.

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -740,18 +524,12 @@ static void SendSysExMsg(unsigned int delta_time, const midi_event_t *event)
-    WriteBuffer((byte *)&native_event, sizeof(native_event_t));
-    WriteBuffer(&event_type, sizeof(byte));
-    WriteBuffer(data, length);
-    WriteBufferPad();
+    SendLongMsg(data, length);
```
`MIDI_EVENT_SYSEX` must be inserted at the front like before otherwise SysEx messages don't work.

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/86f7a213b6d2cc5747721d2ba7995e4aacdd2c63

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -1383,287 +1148,261 @@ static boolean IsRPGLoop(void)
+        if (!song.tracks[i].end_of_track)
         {
-            if (!song.tracks[i].end_of_track)
+            unsigned int time = song.tracks[i].elapsed_time
+                                + MIDI_GetDeltaTime(song.tracks[i].iter);
```
Maybe we should make all the time related variables `uint64_t` to be safe. Same with the `midi_track_t` and `midi_song_t` structures.

rfomin replied:
I don't think it is necessary, `elapsed_time` and `delta_time` are in MIDI ticks, `int32_t` should be sufficient.

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -1854,53 +1566,28 @@ static void I_WIN_UnRegisterSong(void *handle)
-        MidiError("midiStreamClose", mmr);
-    }
-    hMidiStream = NULL;
+    rtmidi_close_port(midiout);
```
There needs to be a check before calling this function to ensure that any pending events are clear from the given OS's midi system. On Windows (not sure about others), it's possible to drop portions of the pitch bend and volume reset messages that are called in `ResetDevice()` because `rtmidi_close_port()` simply calls WinMM's `midiOutClose()` blindly.

ceski-1 replied:
For comparison, before this PR, we carefully waited for `STATE_SHUTDOWN` -> `WaitForSingleObject` (confirms message buffer has been returned) -> `STATE_EXIT` -> reset of shutdown process. Not sure how it would work now.

rfomin replied:
I think we can just add a little delay.

rfomin replied:
Now we check `MIDIERR_STILLPLAYING`, see WINMM `MIDI_SendLongMsg` in midiout.c

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -1383,287 +1148,261 @@ static boolean IsRPGLoop(void)
 
-            case STATE_STOPPED:
-                SetEvent(hStoppedEvent);
+            case STATE_WAITING:
+                {
+                    int64_t remaining_time = wait_time - I_GetTimeUS();
+                    if (remaining_time <= 0)
+                    {
+                        ProcessEvent(event, track);
+                        midi_state = STATE_PLAYING;
+                    }
+                    else if (remaining_time > 1000)
+                    {
+                        sleep = true;
+                    }
+                }
                 break;
```
So if there is 1-1000 us remaining, the loop runs at full speed for that duration? Will that thrash the CPU?

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/d5dcf60e245a3ccff7d3062bc1d583808c19ff09

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -1674,82 +1413,74 @@ static void I_WIN_SetMusicVolume(int volume)
-    LeaveCriticalSection(&CriticalSection);
+    SDL_LockMutex(music_lock);
+    volume_factor = sqrtf((float)volume / 15.0f);
+    if (song.file)
+    {
+        UpdateVolume();
+    }
```
This was changed so it calls directly from the main thread instead of setting a state for later in the midi thread. Now something in i_sound.c or s_sound.c changes the volume when Woof launches (but before `ResetDevice()` is called) and is setting the volume to zero needlessly.

<details><summary>MIDI-OX output</summary>

Before:
```
 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT               
 000016D4   3  --     F0  Buffer:     6 Bytes   System Exclusive      
 SYSX: F0 7E 7F 09 01 F7
[other reset messages]
```

After:
```
 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT               
 00006D42   3  --     B0    07    00    1  ---  CC: Volume            
 00006D42   3  --     B1    07    00    2  ---  CC: Volume            
 00006D42   3  --     B2    07    00    3  ---  CC: Volume            
 00006D42   3  --     B3    07    00    4  ---  CC: Volume            
 00006D42   3  --     B4    07    00    5  ---  CC: Volume            
 00006D42   3  --     B5    07    00    6  ---  CC: Volume            
 00006D42   3  --     B6    07    00    7  ---  CC: Volume            
 00006D42   3  --     B7    07    00    8  ---  CC: Volume            
 00006D42   3  --     B8    07    00    9  ---  CC: Volume            
 00006D43   3  --     B9    07    00   10  ---  CC: Volume            
 00006D43   3  --     BA    07    00   11  ---  CC: Volume            
 00006D43   3  --     BB    07    00   12  ---  CC: Volume            
 00006D43   3  --     BC    07    00   13  ---  CC: Volume            
 00006D43   3  --     BD    07    00   14  ---  CC: Volume            
 00006D43   3  --     BE    07    00   15  ---  CC: Volume            
 00006D43   3  --     BF    07    00   16  ---  CC: Volume   
 00006D43   3  --     F0  Buffer:     6 Bytes   System Exclusive      
 SYSX: F0 7E 7F 09 01 F7
[other reset messages]
```

</details>

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/5b182c593c8aa63069eeb64203a9536d08a9565a

ceski-1 commented on src/i_midimusic.c:
```diff
@@ -1854,53 +1566,28 @@ static void I_WIN_UnRegisterSong(void *handle)
-    }
-    StreamStop();
+    // Send notes/sound off prior to reset to prevent volume spikes.
+    SendNotesSoundOff();
```
`SendNotesSoundOff()` can be removed because it's called above in `I_MID_StopSong()`.

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/ff394d25026903a543b5ad3fe7025b325861060f

ceski-1 commented:
I tried various midi backends across different OSes:
- Windows 10: MS GS Synth misses the first few notes of a song when selecting it from the audio menu.
- Arch Linux: Timidity and a real SC-55 via UM-ONE mk2 aren't initialized properly and sound like pianos even though playing a midi file with `aplaymidi` is fine.
- macOS 14: Native midi no longer appears as an option in audio menu and a real SC-55 has the same piano problem as Linux.

rfomin commented:
> * Arch Linux: Timidity and a real SC-55 via UM-ONE mk2 aren't initialized properly and sound like pianos even though playing a midi file with `aplaymidi` is fine.

My guess is that we need a delay after sysex messages like here: https://github.com/thestk/rtmidi/blob/e8c8791823ca99956adb4faf2b28bb843c7c1f6e/tests/sysextest.cpp#L89

> * macOS 14: Native midi no longer appears as an option in audio menu and a real SC-55 has the same piano problem as Linux.

Too bad, I guess we'll have to open the port ourselves? It seems we need platform specific code anyway. Alternatively, we could try PortMidi.

rfomin commented:
> * Arch Linux: Timidity and a real SC-55 via UM-ONE mk2 aren't initialized properly and sound like pianos even though playing a midi file with `aplaymidi` is fine.

Switching to PortMidi fixes Timidity and Nuked SC-55 reset for me under Linux. It also has sensible device names by default. I haven't tested MacOS.

In general, PortMidi seems to be much more sophisticated than RtMidi.

ceski-1 commented:
Latest commit stutters on music changes and crashes on exit: [musictest2.zip](https://github.com/fabiangreffrath/woof/files/14897805/musictest2.zip)

rfomin commented:
> Latest commit stutters on music changes and crashes on exit: [musictest2.zip](https://github.com/fabiangreffrath/woof/files/14897805/musictest2.zip)

On which platform? No crashes for me on Win10/MS GS Wavetable and Arch Linux/Timidity. How bad is the stutters? I don't see it at 60Hz.

ceski-1 commented:
> On which platform? No crashes for me on Win10/MS GS Wavetable and Arch Linux/Timidity. How bad is the stutters? I don't see it at 60Hz.

Win10/MS GS Synth. Go back and forth a few times. It's also easier to see with bobbing turned off. There's also a frame time spike with MSI Afterburner.

rfomin commented:
> Crashes aren't consistently reproduceable. Race condition in shutdown? See:

We use PortMidi in "real time" mode, without the MidiStream API. See `latency == 0` code paths in `pmwinmm.c`. I haven't seen a crash yet, but I will try to reproduce it.

ceski-1 commented:
> > Crashes aren't consistently reproduceable. Race condition in shutdown? See:
> 
> We use PortMidi in "real time" mode, without the midiStream API. See `latency == 0` code paths in `pmwinmm.c`. I haven't seen a crash yet, but I will try to reproduce it.

It looks like both PortMidi code paths prepare headers and use that callback function.

rfomin commented:
> It looks like both PortMidi code paths prepare headers and use that callback function.

Yes, it seems to use callback for SysEx messages. Eh, maybe the best approach is to combine the best parts of RtMidi and PortMidi and write our own minimal MIDI API backends (only realtime ouput).

rfomin commented:
> Isn't this in a way where we came from with `i_winmusic.c` and `i_macmusic.c`?

The difference is that we will support all `i_winmusic.c` features on other platforms, which is especially interesting with Nuked SC-55 (hardware reset, MIDI complevels, etc).

@ceski-1 I removed PortMidi and implemented `midiout.c`. No stutters/crashes on Win10, timidity/Nuked SC-55 reset works correctly with ALSA. No MacOS yet.

fabiangreffrath commented:
😮

fabiangreffrath commented on src/midiout.c:
```diff
@@ -0,0 +1,400 @@
+
```
Whose code is this?

rfomin replied:
This is mine, but I was looking at RtMidi and PortMidi when I implemented ALSA. This code is trivial API calls, nothing special. The ALSA API is over-complicated, though.

fabiangreffrath commented on vcpkg.json:
```diff
@@ -5,6 +5,7 @@
     "sdl2-net",
     "openal-soft",
     "libsndfile",
+    "portmidi",
```
This shouldn't be necessary anymore.

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/2a947a36ed4f34877186d75f3967e1f9518ea032

rfomin commented:
I have implemented a MIDI backend for macOS, both "DLS Synth" (ala MS GS Wavetable, previously "Native") and CoreMIDI. I still can't get the macOS VM to work on my system, so that's yet to be tested.

ceski-1 commented:
I did some more testing and took notes as I found issues. I'm not sure how much of this will have to be addressed now, in a later PR, or not at all.

<details><summary>Notes</summary>

For macOS, I can't get Woof to see my USB midi interface (it works fine with the macOS "midi test" feature though) or Nuked SC-55. Only the built-in synth (aka DLS), Fluidsynth, and OPL3 appear. I didn't test it further than that yet.

Windows 10 and Arch Linux:

[testwads.zip](https://github.com/fabiangreffrath/woof/files/14986889/testwads.zip)

- dbp37_test.wad:
  - Nuked SC-55 or real SC-55: "MIDI Buff. Full" and piano sounds when selecting device or changing levels too quickly with wipe effect off. I think this is a device limitation.
  - S-YXG50: This fine on Windows. When using Arch (ALSA midi through port + Wine + FSMP), sometimes on launch it plays pianos instead of correct instruments.
  - All of the above might be fixed with `midi_reset_delay > 0` but it seems like a fragile solution.
- musictest.wad:
  - If the .ogg music track is playing and the midi device is switched in the menu, it's not actually switched in-game when you move across the line to change the song. This is easier to hear with OPL3. Maybe you just fixed this as I was typing but I don't have time to verify it right now.
- musictest2.wad: 
  - Moving back and forth creates popping sounds when the music changes. It's much worse when using Arch compared to Windows but it happens for both. Any backend, but Fluidsynth and OPL3 are the worst.

General comments:
- There's a lot of popping in general, not just the above example. Change levels, quit the game, change backends, eventually you'll hear it.
- Linux shows "sndfile/xmp" errors occasionally (try the above wads and switch backends).

</details>

rfomin commented:
> For macOS, I can't get Woof to see my USB midi interface (it works fine with the macOS "midi test" feature though) or Nuked SC-55. Only the built-in synth (aka DLS), Fluidsynth, and OPL3 appear. I didn't test it further than that yet.

It works if you enable "IAC driver". I guess we need to add a virtual MIDI port to make this easier.

> All of the above might be fixed with `midi_reset_delay > 0` but it seems like a fragile solution.

I think adding some delay is not a bad solution (100 ms?). Alternatively, we can add waiting for callback. Currently we use this (midiout.c `MIDI_SendLongMsg`):
```c    
while (MIDIERR_STILLPLAYING
       == midiOutUnprepareHeader(hMidiOut, &hdr, sizeof(MIDIHDR)))
{
    Sleep(1);
}
```

> If the .ogg music track is playing and the midi device is switched in the menu, it's not actually switched in-game when you move across the line to change the song.

Seems to be fixed in https://github.com/fabiangreffrath/woof/pull/1635/commits/e8322d85f22dbcf59a2f5dfc57ebe6e5ca950516

> Moving back and forth creates popping sounds when the music changes. It's much worse when using Arch compared to Windows but it happens for both. Any backend, but Fluidsynth and OPL3 are the worst.

Do we have the same problem with the master branch?

rfomin commented:
> > For macOS, I can't get Woof to see my USB midi interface (it works fine with the macOS "midi test" feature though) or Nuked SC-55. Only the built-in synth (aka DLS), Fluidsynth, and OPL3 appear. I didn't test it further than that yet.
> 
> It works if you enable "IAC driver". I guess we need to add a virtual MIDI port to make this easier.

Turns out that `CFStringGetCStringPtr` is not reliable, it now works on my VM.

rfomin commented:
> All of the above might be fixed with midi_reset_delay > 0 but it seems like a fragile solution.

I think https://github.com/fabiangreffrath/woof/pull/1635/commits/f7a35b0f9036a49306c69c940239a7d490e13dde helps.

> There's a lot of popping in general, not just the above example. Change levels, quit the game, change backends, eventually you'll hear it.

Does https://github.com/fabiangreffrath/woof/pull/1635/commits/0e9ad7e011c9f9d23f3d00a333cbb84990d70465 help?

ceski-1 commented:
> It works if you enable "IAC driver". I guess we need to add a virtual MIDI port to make this easier.
>
> Turns out that `CFStringGetCStringPtr` is not reliable, it now works on my VM.

macOS devices work now but the names aren't correct:

<details><summary>macOS</summary>

![macos_coremidi](https://github.com/fabiangreffrath/woof/assets/56656010/857b0eb5-f777-4a35-b606-73f6e429c850)

They should be something like this (from [SysEx Librarian and MIDI Monitor](https://www.snoize.com/) which are [open source](https://github.com/krevis/MIDIApps) so maybe there's something useful in there):

![macos_sysex_librarian](https://github.com/fabiangreffrath/woof/assets/56656010/ece2ab20-8e6b-4915-9ca3-f328778a6622)

</details>

ceski-1 commented:
> Do we have the same problem with the master branch?
>
> I think [f7a35b0](https://github.com/fabiangreffrath/woof/commit/f7a35b0f9036a49306c69c940239a7d490e13dde) helps.
>
> Does [0e9ad7e](https://github.com/fabiangreffrath/woof/commit/0e9ad7e011c9f9d23f3d00a333cbb84990d70465) help?

Those don't help and I confirmed that the master branch has some of the same problems as this PR.

<details><summary>Notes</summary>

Both master branch and this PR:

- Any OS:
  - Fluidsynth and OPL3: both have popping.
  - Nuked SC-55: similar issues as real hardware, meaning `midi_reset_delay > 0` or notes stutter, "MIDI Buff. Full", wrong instruments (pianos).
- Windows:
  - MS GS Synth: has popping.

This PR only:

- Any OS:
  - Fluidsynth and OPL3: first few notes are cut off for every song. The master branch has this issue but only with Linux.
  - Song transitions are slow and sometimes overlap (e.g. going from map to intermission screen, musictest2.wad, CRASH_test.wad).
- Linux:
  - Timidity needs `midi_reset_delay > 0` and has popping.
  - S-YXG50 (Wine): needs `midi_reset_delay > 0` or pianos play.
- macOS:
  - Real SC-55: only plays pianos and delays don't help.

These are fine:

- Windows:
  - VirtualMidiSynth
  - X-Fi Titanium
  - SCVA
  - S-YXG50
- macOS:
  - DLS Synth

</details>

rfomin commented:
> Any OS:
> * Fluidsynth and OPL3: both have popping.

I think it's an OpenAL problem. We can add some sort of "fading". To be honest, I can't hear it (although my hearing is not good).

> Windows:
> * MS GS Synth: has popping.

Maybe I don't understand what "popping" means. Isn't it about the transition between tracks? MS GS Synth sounds less "harsh" to me.

> This PR only:
> Any OS:
> * Fluidsynth and OPL3: first few notes are cut off for every song. The master branch has this issue but only with Linux.
> * Song transitions are slow and sometimes overlap (e.g. going from map to intermission screen, musictest2.wad, CRASH_test.wad).

I can't replicate this. I didn't change Fluidsynth and OPL3 in this PR, just a little i_oalmusic.c tweaking.

> Linux:
> * Timidity needs midi_reset_delay > 0 and has popping.

Timidity works fine for me. I'm using just ALSA (no PulseAudio etc.) on Arch and run it with `timidity -iA` command.

> macOS:
> Real SC-55: only plays pianos and delays don't help.

Nuked SC-55 works? (It works for me).

> macOS devices work now but the names aren't correct:

Looks like Unicode issue. Could you try this patch:

```diff
diff --git a/src/midiout.c b/src/midiout.c
index 008ca37..abc2a5f 100644
--- a/src/midiout.c
+++ b/src/midiout.c
@@ -434,7 +434,7 @@ static void Init(void)
         {
             int len = CFStringGetLength(name);
             char *s = malloc(len);
-            if (!CFStringGetCString(name, s, len, kCFStringEncodingASCII))
+            if (!CFStringGetCString(name, s, len, kCFStringEncodingMacRoman))
             {
                 int k;
                 for (k = 0; k < len; ++k)
```

rfomin commented:
> macOS devices work now but the names aren't correct:

I think it's Unicode issue. Does https://github.com/fabiangreffrath/woof/pull/1635/commits/ebbdd54966761dac60ee1289763703324d00ae8e help?

RtMidi has this: https://github.com/thestk/rtmidi/blob/e8c8791823ca99956adb4faf2b28bb843c7c1f6e/RtMidi.cpp#L1317 We can try it.

ceski-1 commented:
> Maybe I don't understand what "popping" means. Isn't it about the transition between tracks? MS GS Synth sounds less "harsh" to me.

It's a sharp popping sound that sometimes occurs 
when stopping a song. So that means selecting a device in the menu or quitting the game. Other backends struggle with musinfo changes, but MS GS Synth seemed fine with those for some reason.

> I can't replicate this. I didn't change Fluidsynth and OPL3 in this PR, just a little i_oalmusic.c tweaking.

A good song to test is doom2 map07 since it has a note right at the start.

> Nuked SC-55 works? (It works for me).

Yes, see near the top of the notes for the caveats. If Nuked SC-55 becomes popular it will require some care due to the limitations of the original hardware it reproduces. I'll take a second look at a real SC-55 under macOS tomorrow. Maybe my midi interface is misconfigured.

ceski-1 commented:
> It's a sharp popping sound that sometimes occurs
when **stopping a song**. So that means selecting a device in the menu or quitting the game. Other backends struggle with musinfo changes, but MS GS Synth seemed fine with those for some reason.

Looking at my comment again, I think it's when the _device_ is stopped and doesn't have time to fade out. So not necessarily stopping a song.

ceski-1 commented:
> I think it's Unicode issue. Does [ebbdd54](https://github.com/fabiangreffrath/woof/commit/ebbdd54966761dac60ee1289763703324d00ae8e) help?

Yes, the names show up now. It's "Bus 1" for IAC Driver Bus 1 which seems a bit short, but I don't have a strong opinion.

rfomin commented:
> > I can't replicate this. I didn't change Fluidsynth and OPL3 in this PR, just a little i_oalmusic.c tweaking.
> 
> A good song to test is doom2 map07 since it has a note right at the start.

https://github.com/fabiangreffrath/woof/pull/1635/commits/2f2f17b3a14d39389b19ecb93d9a323f998dfc8b fixed it for me.

rfomin commented:
@ceski-1 I reverted 10a48c9, as it didn't help. It looks like OpenAL-Soft does some fading when calling `alSourceStop` (discussion here kcat/openal-soft#273), but I'm not sure if this is in the current version. I think the "popping" issues are beyond the scope of this PR.

I believe this PR is complete, what do you think?

ceski-1 commented:
> I believe this PR is complete, what do you think?

I think so. Like you said, the other issues are out of scope or not a high priority in my opinion.
